In [10]:
def run_timing():
    times = []
    while True:
        try:
            run_time = input("Combien de temps (en minutes) ça vous prend pour courir 10km?")
            if run_time == "":
                break
            times.append(float(run_time))
        except ValueError: print("invalide")
    if times: # Vérifie que la liste n'est pas vide pour ne pas diviser par 0
        moyenne = sum(times) / len(times)
        print(f"Vous avez couru en moyenne {moyenne} kms sur {len(times)} courses")
    else:
        print("Vous n'avez pas couru")





In [12]:
run_timing()

Vous avez couru en moyenne 9.0 kms sur 3 courses


In [4]:
s = 0.1 + 0.2
>>> print(s)

0.30000000000000004


# Garder le contrôle avec f-strings

On peut voir que la réponse n'est pas bonne. Pourquoi?

L’ordinateur n’utilise que des 0 et 1 (base 2).
Les nombres à virgule (comme 0.1 ou 0.2) doivent être représentés sous forme binaire.

0.1_10 (qui est 1/10), en binaire c’est une décimale infinie récurrente : 0.00011001100110011...
donc 0.1 + 0.2 = 0.30000000000000004 ce qui est une approximation

on peut utiliser les f-strings pour contourner le problème et limiter le nombre de chiffres après la virgule. À noter que cela n'arrondit pas le nombre au sens mathématique "par défaut", ça le formate à 2 chiffres après la virgule.

**Le résultat affiché est arrondi, pas le nombre lui-même**

In [8]:
print(f'{s:.2f}')

0.30


In [9]:
def run_timingV2():
    times = []
    while True:
        try:
            run_time = input("Combien de temps (en minutes) ça vous prend pour courir 10km?")
            if run_time == "":
                break
            times.append(float(run_time))
        except ValueError: print("invalide")
    if times: # Vérifie que la liste n'est pas vide pour ne pas diviser par 0
        moyenne = sum(times) / len(times)
        print(f"Vous avez couru en moyenne {moyenne:.2f} kms sur {len(times)} courses")
    else:
        print("Vous n'avez pas couru")

In [10]:
run_timingV2()

Vous avez couru en moyenne 0.15 kms sur 2 courses
